sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -327,12 +342,42 @@ DataSource buildDatasource() throws Exception{
         if (getConnectionProperties() != null && getConnectionProperties().get() != null) {
           basicDataSource.setConnectionProperties(getConnectionProperties().get());
         }
-        return basicDataSource;
+        current = basicDataSource;
       }
+
+      // wrapping the datasource as a pooling datasource
+      DataSourceConnectionFactory connectionFactory = new DataSourceConnectionFactory(current);
+      PoolableConnectionFactory poolableConnectionFactory =
+              new PoolableConnectionFactory(connectionFactory, null);
+      GenericObjectPoolConfig poolConfig = new GenericObjectPoolConfig();
+      poolConfig.setMaxTotal(getPoolMaxTotal());
+      poolConfig.setBlockWhenExhausted(true);
```

Comment:
Almost all of these parameters are also irrelevant because there will be at most 1 connection - please keep only the necessary ones. I think really you only need maxTotal=1, minIdle=0.

Reply:
And you also need to configure eviction as per https://stackoverflow.com/questions/38685092/dbcp2-when-are-idle-connections-removed-from-the-pool - otherwise, with minIdle=0, the idle connections will be closed immediately on release rather than after an idle timeout.